### PR TITLE
Sharing modules with `odra(using)` attribute

### PR DESCRIPTION
**Pull request by kpob**
* change StaticInstance::instance signature
* remove composer
* simplify Node trait
* remove skip_instance attribute
* handle `using` attribute

### CODE REVIEW

**Review by kubaplas — APPROVED** [review]

File: lang/codegen/src/generator/module_item/node.rs [hunk 1]
```diff
@@ -28,34 +28,48 @@ impl GenerateCode for Node<'_> {
...
+            .delegated_fields
+            .iter()
             .map(|field| {
-                let ty = &field.ty;
-                let ident = field.ident.as_ref().unwrap();
+                let ty = &field.field.ty;
+                let ident = field.field.ident.as_ref().unwrap().to_string();
+                let a = field.delegated_fields.iter().map(|f| quote!(#f)).collect::<Punctuated<TokenStream, Token![,]>>();
+
+                // let key = &String::from("shared#value");
```
Comment (kubaplas) [inline lang/codegen/src/generator/module_item/node.rs]: remove

File: odra-casper/codegen/src/lib.rs [hunk 2]
```diff
@@ -97,8 +97,8 @@ macro_rules! gen_contract {
                     panic!("Event {} can't have Type::Any struct in it.", &event.ident);
                 }
             }
-            let keys = <$contract as odra::types::contract_def::Node>::keys();
-            let keys_count = <$contract as odra::types::contract_def::Node>::count();
+            let keys = <$contract as odra::types::contract_def::Node>::_keys();
```
Comment (kubaplas) [inline odra-casper/codegen/src/lib.rs]: __keys

**Review by kubaplas — APPROVED** [review]

File: lang/codegen/src/generator/module_impl/deploy/deployer_casper_livenet.rs [hunk 3]
```diff
@@ -49,7 +49,12 @@ fn build_entrypoint_call(entrypoint: &Method, struct_ident: &Ident) -> TokenStre
     let arg_names = args_to_arg_names_stream(&entrypoint.args);
     quote! {
         entrypoints.insert(#name, (#arg_names, |name, args| {
-            let mut instance = <#struct_ident as odra::DynamicInstance>::instance(&[]);
+            let keys = <#struct_ident as odra::types::contract_def::Node>::__keys();
+            let keys = keys
+                .iter()
+                .map(String::as_str)
+                .collect::<Vec<_>>();
+            let (mut instance, _) = <#struct_ident as odra::StaticInstance>::instance(keys.as_slice());
```
Comment (zie1ony) [inline lang/codegen/src/generator/module_impl/deploy/deployer_casper_livenet.rs:57]: Shouldn't you check that the second elem of tuple is empty?

File: lang/codegen/src/generator/module_impl/deploy/deployer_mock_vm.rs [hunk 4]
```diff
@@ -137,7 +142,12 @@ fn build_entrypoints_calls(methods: &[&Method], struct_ident: &Ident) -> TokenSt
                 entrypoints.insert(#name, (#arg_names, |name, args| {
                     #reentrancy_check
                     #attached_value_check
-                    let mut instance = <#struct_ident as odra::DynamicInstance>::instance(&[]);
+                    let keys = <#struct_ident as odra::types::contract_def::Node>::__keys();
+                    let keys = keys
+                        .iter()
+                        .map(String::as_str)
+                        .collect::<Vec<_>>();
+                    let (mut instance, _) = <#struct_ident as odra::StaticInstance>::instance(keys.as_slice());
```
Comment (zie1ony) [inline lang/codegen/src/generator/module_impl/deploy/deployer_mock_vm.rs:150]: Same here.

File: lang/codegen/src/generator/module_impl/deploy/deployer_mock_vm.rs [hunk 5]
```diff
@@ -88,7 +88,12 @@ fn build_constructor(
                 stringify!(#constructor_ident).to_string(),
                 &args,
                 |name, args| {
-                    let mut instance = <#struct_ident as odra::DynamicInstance>::instance(&[]);
+                    let keys = <#struct_ident as odra::types::contract_def::Node>::__keys();
+                    let keys = keys
+                        .iter()
+                        .map(String::as_str)
+                        .collect::<Vec<_>>();
+                    let (mut instance, _) = <#struct_ident as odra::StaticInstance>::instance(keys.as_slice());
```
Comment (zie1ony) [inline lang/codegen/src/generator/module_impl/deploy/deployer_mock_vm.rs:96]: Same here

**Review by zie1ony — APPROVED** [review]

**Pull request merged**